cmd/zrepl/mgmt_conn.c

```diff
@@ -0,0 +1,907 @@
+		conn->conn_state = CS_READ_VERSION;
+		break;
+	case CS_READ_VERSION:
+		vers = *((uint16_t *)conn->conn_buf);
+		kmem_free(conn->conn_buf, sizeof (uint16_t));
+		if (vers != REPLICA_VERSION) {
+			fprintf(stderr, "invalid replica protocol version %d\n",
+			    vers);
+			rc = reply_error(conn, ZVOL_OP_STATUS_VERSION_MISMATCH,
+			    0, 0, CS_CLOSE);
+		} else {
+			DBGCONN(conn, "Reading header..");
+			hdrp = kmem_zalloc(sizeof (*hdrp), KM_SLEEP);
+			hdrp->version = vers;
+			conn->conn_buf = hdrp;
+			conn->conn_bufsiz = sizeof (*hdrp);
+			conn->conn_procn = sizeof (uint16_t); // skip version
+			conn->conn_state = CS_READ_HEADER;
+		}
+		break;
+	case CS_READ_HEADER:
+		hdrp = conn->conn_buf;
+		conn->conn_hdr = hdrp;
+		if (hdrp->len > 0) {
+			DBGCONN(conn, "Reading payload (%lu bytes)..",
+			    hdrp->len);
+			conn->conn_buf = kmem_zalloc(hdrp->len, KM_SLEEP);
+			conn->conn_bufsiz = hdrp->len;
+			conn->conn_procn = 0;
+			conn->conn_state = CS_READ_PAYLOAD;
+		} else {
+			conn->conn_buf = NULL;
+			conn->conn_bufsiz = 0;
+			rc = process_message(conn);
+		}
+		break;
+	case CS_READ_PAYLOAD:
+		rc = process_message(conn);
+		break;
+	default:
+		ASSERT(0);
+		/* Fall-through */
+	case CS_CLOSE:
+		rc = close_conn(conn);
+		break;
+	}
+
+	return (rc);
+}
+
+/*
+ * One thread to serve all management connections operating in non-blocking
+ * event driven style.
+ */
+void
+uzfs_zvol_mgmt_thread(void *arg)
+{
+	char			*buf;
+	uzfs_mgmt_conn_t	*conn;
+	struct epoll_event	ev, events[MAX_EVENTS];
+	int			nfds, i, rc;
+	boolean_t		do_scan;
+
+	mutex_init(&conn_list_mtx, NULL, MUTEX_DEFAULT, NULL);
+
+	mgmt_eventfd = eventfd(0, EFD_NONBLOCK);
+	if (mgmt_eventfd < 0) {
+		perror("eventfd");
+		zk_thread_exit();
+		return;
+	}
+	epollfd = epoll_create1(0);
+	if (epollfd < 0) {
+		perror("epoll_create1");
+		zk_thread_exit();
+		return;
+	}
+	ev.events = EPOLLIN;
+	ev.data.ptr = NULL;
+	if (epoll_ctl(epollfd, EPOLL_CTL_ADD, mgmt_eventfd, &ev) == -1) {
+		perror("epoll_ctl");
+		zk_thread_exit();
+	}
+
+	prctl(PR_SET_NAME, "mgmt_conn", 0, 0, 0);
+
+	/*
+	 * The only reason to break from this loop is a failure to update FDs
+	 * in poll set. In that case we cannot guarantee consistent state.
+	 * Any other failure should be handled gracefully.
+	 */
+	while (1) {
+		do_scan = B_FALSE;
+		nfds = epoll_wait(epollfd, events, MAX_EVENTS,
+		    1000 * RECONNECT_DELAY / 2);
+		if (nfds == -1) {
+			if (errno == EINTR)
+				continue;
+			perror("epoll_wait");
+			goto exit;
+		}
+
+		for (i = 0; i < nfds; i++) {
+			conn = events[i].data.ptr;
+
+			/*
+			 * data.ptr is null only for eventfd. In that case
+			 * zinfo was created or deleted -> scan the list.
+			 */
+			if (conn == NULL) {
+				uint64_t value;
+
+				do_scan = B_TRUE;
+				/* consume the event */
+				rc = read(mgmt_eventfd, &value, sizeof (value));
+				ASSERT3P(rc, ==, sizeof (value));
+				continue;
+			}
+
+			if (events[i].events & EPOLLERR) {
```

Comment:
better to change this as !EPOLLOUT and !EPOLLIN

Reply:
There are certain flags which are always set by epoll_wait() (when there is a reason to set them) even if user did not request those flags (subscribed to those events) and those are EPOLLERR and EPOLLHUP. EPOLLHUP means the peer closed the connection from his side. We handle that case by EPOLLIN and read() zero bytes so no need to detect this flag. Now if we consider everything what is not EPOLLIN and EPOLLOUT to be an error, that could have strange consequences depending on if EPOLLIN is set in case of EPOLLHUP. IMHO using explicit EPOLLERR as more simple and clear. But maybe I haven't considered some cases where !EPOLLIN && !POLLOUT would be better option. Is it common to use it in that way in other open source projects which use epoll()?